File: PeakInRotatedSortedArr.java
// Time Complexity :O(log n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


/**
 *1. Find mid of an array. If mid is greater than mid+1 search the left part of array.
 *2. else if mid is smaller than mid+1 search right part of array
 *3. -1 and n index is least int so in worst case it will return 0 or n-1 index.
 *4. missed condition m[i] != m[i+1]
 *
 */
class Solution {
    public int findPeakElement(int[] nums) {
    	 int l = 0, r = nums.length - 1;
         while (l <= r) {
             int mid = l + (r-l) / 2;
             if (mid!=nums.length-1 && nums[mid] > nums[mid + 1])
                 r = mid-1;
             else if(mid!=nums.length-1)
                 l = mid + 1;
             else
            	 return l;
         }
         return l;
    }
}
public class PeakInRotatedSortedArr {
	
	public static void main(String[] args) {
		System.out.println(new Solution().findPeakElement(new int[]{11,9,3,8,9,11}));
	}

}

File: peakInRotatedSortedArr.cpp
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int findPeakElement(vector<int>& nums) 
    {
        int n = nums.size()-1;
        if(nums.size() ==1)
        {
            return 0;
        }
        else if (nums[0]>nums[1])
        {
            return 0;
        }
        else if(nums[n] > nums[n-1])
        {
            return n;
        }
        int left = 1;
        int right = nums.size()-2;
        int mid = 0;
        while(left<=right)
        {
            mid = (left+right)/2;

            if(nums[mid]>nums[mid+1] && nums[mid]>nums[mid-1])
            {
                return mid;
            }
            if(nums[mid]<nums[mid-1])
            {
                right = mid -1;
            }
            else
            {
                left = mid + 1; 
            }
        }
        return -1;
    }
};